pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages. 10 is a
+// reasonable default that balances usefulness with performance impact. It isn't
+// expected that this limit will normally be hit.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// maybeRegister checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The invariant is
+// that numRegistered <= numAllowed.  This method will return true if we can
+// keep the invariant and added one to the number registered, otherwise it will
+// return false.
+func (r *RaftTracer) maybeRegister() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegistered := r.numRegistered.Load()
+
+	// The maximum number of traces has been reached. We don't register this
+	// trace and return false.
+	if numRegistered == numAllowed {
+		return false
+	}
+
+	// This can only happen if numAllowed has changed. If this happens flush all
+	// our current traces and don't register this request.
+	for numRegistered > numAllowed {
+		r.FlushAll()
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	return r.numRegistered.CompareAndSwap(numRegistered, numRegistered+1)
+}
+
+// RegisterRemote is used to register a remote trace. This is used when we
+// receive a raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(e kvserverpb.TracedEntry) {
+	if !r.maybeRegister() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	ctx := logtags.AddTag(r.ctx, "id", tracingID(e.TraceID, e.SpanID))
+	t := traceValue{traceID: e.TraceID, spanID: e.SpanID, ctx: ctx, baseCtx: nil}
+	t.logf(1, "start trace for index %d from remote", e.Index)
+	r.m.Store(e.Index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if !r.maybeRegister() {
+		return
+	}
+	span := tracing.SpanFromContext(baseCtx.Context())
```

Comment:
Should we avoid the `Context()` callback-style API, and just pass the `span` or `Context` to `Register` func directly? It can be fetched from the proposal context on the caller side.

Reply:
The reason for this API is because the context changes under us, and we need to stop writing to it once the caller change it. I added a comment describing this behavior.

Reply:
The caller does the same `SpanFromContext` call, so we end up doing this twice. We could just pass the context in instead, to avoid the double work.

Reply:
I changed to move calling SpanFromContext into this method only. There are a few cases where it doesn't always register. This also removes any risk of the context changing between these calls (although in practice this couldn't happen).